Update README and introduce Developer instructions (#78)
impermeable/coq-waterproof: Developer-instructions.md

# Developer instructions

## Recommended setup

Our recommended setup for working on the library is by using vscode on Linux, MacOS, or WSL, with
the coq-lsp and ocaml extensions. We will now describe more extensively how to achieve this setup.

### Installation of opam

A more extensive description can be found [here](https://ocaml.org/docs/installing-ocaml).
This is a short description.

On some versions of Linux and WSL, you can install opam with
```
apt-get install opam
```
On MacOs, you can either use homebrew
```
brew install gpatch
brew install opam
```
or macports
```
port install opam
```

### Creating the desired opam environment: simplified version
When we develop, we need to be aware of the version of Coq we develop for.

> Most development is done against the main branch, and then the following setup will suffice.

Reviewer: Maybe
```suggestion
Most development is done based off the main branch, and then the following setup will suffice.
```
"against" made me think it was the main branch of something else (oCaml or Coq)

```
opam init
eval $(opam env)

> opam install coq-lsp.2.0.0+8.17

Reviewer: How often does this version change? If it changes often maybe note that above this command, if there's a way we can reliably install some latest version, maybe give that command instead. Maybe the edit below is enough?

Author: I'll go with your suggestion below. We may indeed want to think about how to get all version information up to date.

opam install ocaml-lsp-server
```

> 

Reviewer: ```suggestion
replacing 8.17 with the desired version of Coq

```

### Creating the desired opam environment: advanced version for supporting multiple versions of Coq

When we bring changes into different versions of Coq (for instance for the branches 8.18, 8.19...)
it is convenient to have multiple opam environement, i.e. switches, available. If you prefer to start
with such a setup anyway, you can run
```
opam init --bare
``` 
Regardless of how you initialized opam, you can now add a new switch with
```
opam switch create your_preferred_switch_name ocaml-base-compiler.4.14.1
```
Next you can install the background libraries again
```
opam install coq-lsp.2.0.0+8.17
opam install ocaml-lsp-server
```

> replacing 8.17 with the desired version of Coq

Reviewer: ```suggestion
again replacing 8.17 with the desired version of Coq
```

### Caveat: locally compiling the coq-master branch

The above only works partially when checking work for `coq-master`,  which is a very important
branch for us because it is checked in Coq's CI. To just check locally whether the code works on the 
`coq-master` branch, we can setup
```
opam switch create your_preferred_switch_name ocaml-base-compiler.4.14.1
```
Now, we need to add a remote repository to get the development version of coq from.
```
opam repo add coq-dev https://coq.inria.fr/opam/core-dev
opam update
```
Now we can install
```
opam install coq.dev
```

## Building with dune

Our go-to way for compiling `coq-waterproof` while working on the library is by using dune. To do so, from the project's
root (i.e. inside the directory `...some-path.../coq-waterproof/`), run
```
dune build -p coq-waterproof
```
To then install the library so it is available from other projects, you can run
```
dune install -p coq-waterproof
```

## Testing

We aim to test every feature that we implement. Currently we test almost
all functionality present in `.v` files. Note for instance that tactics
such as `Choose`, `Assume`, both have a file that defines them, and a file

> that tests them.

Reviewer: How do I run the tests?

Author: I'll add an explanation

## Running external tests

If one has a folder `waterproof-exercises` at the same level as the folder
`coq-waterproof`, one can test coq-waterproof against all the files 
(typically exercise sheets) in `waterproof-exercises` by executing

```
dune build -p coq-waterproof @runtest
```

## Building with make

In principle, we use **dune** to build `coq-waterproof`, but we try to 
still support building with **make** as well. Before creating a PR, it is
recommended to check that this really works. For this, both `_CoqProjectForMake`
and `waterproof.mlpack` need to be in order, i.e. they need to contain
the correct files, and in particular `waterproof.mlpack` needs the files 
to be in the right order.

TODO: include a test that checks that all .v, .ml files are actually included in the
`_CoqProjectForMake` file.

To then compile and install, run from the project's root directory:
```
make
make install
```

## Building with make (legacy method)

In previous versions of coq-waterproof, before running make, one needed to run,
from the project's root directory,

```
autoreconf -i -s
./configure
```
after which one could run
```
make
make install
```

> One needs to have the `autoconf` ocaml libary installed. One can install this with

Reviewer: Or equivalent commands for non-Debian environments.

```
sudo apt-get update
sudo apt-get install autoconf
```

> 

Reviewer: ```suggestion
or equivalent for non-Debian/Ubuntu based environments

```

## Making Ocaml functions available from Ltac2: Using the foreign function interface (ffi)

One can make Ocaml functions available for use in Ltac2 by using the
foreign function interface (ffi). For now, we gather all such 
definitions in `Wp_ffi.ml`.

Suppose one has an Ocaml function `f : a -> b -> c -> d` where `a, b, c, d` are types, and one would like to make this function available as an Ltac2 tactic.

### Make a tactic

First of all, only tactics can pass the ffi, so one needs to convert `f` into
something that outputs a tactic. An element of type `'a` can be converted into an `'a`-valued tactic, ie.e. an element of type `'a tactic` by using `tclUNIT`:

Consider this example:
```
let my_unit_tactic : unit tactic = tclUNIT ()
```

```
let my_bool_tactic : bool tactic = tclUNIT true
```

Hence, we can convert `f` into something that outputs a tactic by

```
let my_f_tactic : a -> b -> c -> d tactic =
  fun input_a input_b input_c -> tclUNIT (f input_a input_b input_c)
```

### Make a `valexpr` tactic

> Yet datatypes in Ocaml and in Ltac2 need to be translated to each other

Reviewer: ```suggestion
Datatypes in Ocaml and in Ltac2 need to be translated to each other
```
If this was into contrast to the part before I don't really understand it, so alternative to removing yet is explaining the contrast in a bit more depth.

> by using an intermediate datatype: namely `valexpr`. This also means that
> not any tactic can pass the ffi, but only `valexpr`-valued tactics.

Reviewer: ```suggestion
by using an intermediate datatype: namely `valexpr`. This also means that
only `valexpr`-valued tactics can pass the ffi.
```
For clarity?

```
apt-get install opam
```
On MacOs, you can either use homebrew
```
brew install gpatch
brew install opam
```
or macports
```
port install opam
```

### Creating the desired opam environment: simplified version
When we develop, we need to be aware of the version of Coq we develop for.
Most development is done based off the main branch, and then the following setup will suffice.

```
opam init
eval $(opam env)
opam install coq-lsp.2.0.0+8.17
opam install ocaml-lsp-server
```
replacing 8.17 with the desired version of Coq, or more generally, replacing 
2.0.0+8.17 by the desired version of coq-lsp.

### Creating the desired opam environment: advanced version for supporting multiple versions of Coq

When we bring changes into different versions of Coq (for instance for the branches 8.18, 8.19...)
it is convenient to have multiple opam environement, i.e. switches, available. If you prefer to start
with such a setup anyway, you can run
```
opam init --bare
``` 
Regardless of how you initialized opam, you can now add a new switch with
```
opam switch create your_preferred_switch_name ocaml-base-compiler.4.14.1
```
Next you can install the background libraries again
```

> opam install coq-lsp.2.0.0+8.17

Reviewer 2: Think you folks got a typo here, do you mean `coq-lsp.0.2.0+8.17` ?

Author: Oops! You're right, thanks!
